TY: derive Sized trait impl

Undin opened the pull request:
Part of #1751

Undin commented on src/main/kotlin/org/rust/lang/core/types/ty/Ty.kt:
```diff
@@ -85,3 +88,30 @@ fun mergeFlags(element: BoundElement<*>): TypeFlags =
 fun mergeFlags(tys: List<Ty>): TypeFlags =
     tys.fold(0) { a, b -> a or b.flags }
+
+tailrec fun Ty.isSized(): Boolean {
```
I don't really like that this function is public but I haven't come up with how to make it private because I use it in two places

matklad replied:
That's fine I think.

Undin commented on src/main/kotlin/org/rust/lang/core/types/ty/TyTypeParameter.kt:
```diff
@@ -58,15 +61,22 @@ class TyTypeParameter private constructor(
+        fun self(item: RsTraitOrImpl): TyTypeParameter {
+            val isSized = when (item) {
+                // TODO: support trait bounds
+                is RsTraitItem -> false
```
I don't know if we really want to check trait bounds here because we get all bounds in `ImplLookup#rawFindImplsAndTraits` and it brings `Sized` trait in result set so related test passes because of this hack.

vlad20012 replied:
All this code is a set of hacks and will be dismantled soon, so everything is OK =)

matklad commented on src/main/kotlin/org/rust/lang/core/psi/ext/RsTypeParameter.kt:
```diff
@@ -24,6 +24,13 @@ val RsTypeParameter.bounds: List<RsPolybound> get() {
+val RsTypeParameter.isSized: Boolean get() {
+    return stub?.isSized ?: bounds.none {
+        if (it.q == null) return@none false
+        it.bound.traitRef?.resolveToTrait?.isSizedTrait ?: false
```
For dealing with nullable booleans, ` == true` is more idiomatic: https://github.com/JetBrains/kotlin-web-site/blob/465783bf395acc5d34f089291eca594ef3d7d4ce/pages/docs/reference/coding-conventions.md#using-nullable-boolean-values-in-conditions

matklad commented on src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt:
```diff
@@ -91,24 +93,28 @@ class ImplLookup(
-            )
         }
+        implsAndTraits += findImplicitImpls(ty)
+        return implsAndTraits
```
👍

matklad replied:
In general, I think we should switch our style from recklessly concatenating read only collections to  the pattern of creating a mutable collection, filling it with values and then freezing it. There are helper functions like `buildList` which could help here, though in this particular case just creating a mutable set seems clearer.

matklad commented on src/main/kotlin/org/rust/lang/core/types/ty/Ty.kt:
```diff
@@ -85,3 +88,30 @@ fun mergeFlags(element: BoundElement<*>): TypeFlags =
+        }
+
+        is TyTuple -> types.last().isSized()
+        else -> false
```
Is `TyNever` unhanded? 

I think we should return `true` by default here, because you have to opt-out of sizedness usually.

Undin replied:
> Is TyNever unhanded?

`TyNever` is primitive type so it is sized too)

matklad commented on src/main/kotlin/org/rust/lang/core/types/ty/TyTypeParameter.kt:
```diff
@@ -80,10 +90,17 @@ class TyTypeParameter private constructor(
                 AssociatedType(type, trait, target),
-                { emptyList() } )
+                // TODO: support associated type bounds
+                true,
```
Bare `true` looks not very readable on GitHub, let's add `isSized = true` for readability?

matklad commented:
LGTM!

r? @vlad20012

Undin commented:
Fixed comments

vlad20012 commented on src/test/kotlin/org/rust/lang/core/type/RsImplicitTraitsTest.kt:
```diff
@@ -0,0 +1,165 @@
+    """)
+
+    fun `test type parameter with ?Sized bound is not Sized`() = doTest("""
+        fn foo<T: ?Sized>() -> Box<T> { unimplemented!() }
```
Let's add a test where `T: Sized` explicitly: `fn foo<T: Sized>`. It allows to make sure that we handle `?` correctly

Undin replied:
Done

vlad20012 commented:
Thanks a lot!
bors r+

vlad20012 commented:
bors r-

vlad20012 commented:
Hmm, I just found what's done wrong. When I wrote the issue, I mentioned `findImplsAndTraits`, and that was true at that time. But now I even don't sure that we need to change anything in `findImplsAndTraits`. The main place we now need Sized/Send/Sync auto-derive is `assembleCandidates`. And `isSized`/`isCopy` should be implemented like `select(TraitRef(ty, sizedTrait .withSubst()))`

Undin commented:
Ok, I'll fix it a little bit later)

Undin commented:
@vlad20012 fixed

vlad20012 commented on src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt:
```diff
@@ -91,24 +93,27 @@ class ImplLookup(
-        return when (ty) {
-            is TyTypeParameter -> ty.getTraitBoundsTransitively().map { it.element }
-            is TyTraitObject -> BoundElement(ty.trait).flattenHierarchy.map { it.element }
+    private fun rawFindImplsAndTraits(ty: Ty): Set<RsTraitOrImpl> {
```
As I see, nothing changed here except we now use `Set` instead of `List`. What are the reasons for this? I'm worried about this also because the returned collection always stores to the cache, so it is long-leaving object, and HashSet consumes much more memoty than plain ArrayList.

matklad replied:
Semantically, we want a `Set` of impls here. Is there a compact array-based readonly set implementation somewhere?

vlad20012 replied:
Furthermore, `mutableSetOf` used there, so actually it's the `LinkedHashSet` =)

@matklad We don't have something like `TObjectHashSet` (but have in `Koloboke` collections! which we don't use...). Anyway I think that semantic meaning is secondary in as hot place as this.

matklad replied:
Hm, but can we have both proper semantic & good performance? 

```
class ArraySet<T>(private val data: Array<T>) : Set<T> {
    override val size: Int get() = data.size
    override fun contains(element: T): Boolean = data.contains(element)
    override fun containsAll(elements: Collection<T>): Boolean = elements.all { data.contains(it) }
    override fun isEmpty(): Boolean = data.isEmpty()
    override fun iterator(): Iterator<T> = data.iterator()

    companion object {
        inline fun <reified T> fromSet(s: Set<T>): ArraySet<T> = ArraySet(s.toTypedArray())
    }
}
```

> so actually it's the LinkedHashSet =)

Being deterministic is a nice property, for tests especially. So, would constructing a LinkedHashSet and then converting it to an ArraySet be OK performance-wise?

vlad20012 replied:
@matklad As I know, java's `Set` guarantees O(1) for `contains`... And it looks rather like a hack than more beautiful solution =)

matklad replied:
>As I know, java's Set must provide O(1) for contains.

Even `TreeSet`? :trollface:

vlad20012 replied:
Hmm, but not O(n) anyway! =D

vlad20012 replied:
About @Undin's description for `Set` in the comment below:
Hmm, it's reasonable. It would be cool to add some tests for method lookup with `fn foo<T: Trait>(t: T) where T: Trait { ... }`. Without a `Set` we will run into ambiguity?

vlad20012 commented on src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt:
```diff
@@ -219,9 +224,10 @@ class ImplLookup(
     }
 
-    private fun assembleCandidates(ref: TraitRef): List<SelectionCandidate> = buildList {
+    private fun assembleCandidates(ref: TraitRef): Set<SelectionCandidate> = buildSet {
```
Same question about `Set`, but here, I guess, the point is that we can somehow run into multiple implementations of `Sized` trait (but I don't understand how). Anyway, think it's ok for now because I going to optimize this method (i.e. why we need to call the costly `assembleImplCandidates` for type parameters?), and we can handle `Sized` trait as a special case in the same manner (i.e. why we need to call the costly `assembleImplCandidates` if we selecting `Sized` trait that can't be implemented manually?).

Undin replied:
Very stupid case is
```rust
fn foo<T: Trait>(t: T) where T: Trait { ... }
```
It looks like valid rust function for any `Trait`.

My case is
```rust
fn foo<T: Sized>(t: T) { ... }
```
In this case `Sized` trait comes from explicit type bound and from `assembleImplicitCandidates`. Of course, I can check this case manually but I think `Set`
 can solve both problems more elegantly.

Undin replied:
Hm, if we want to have special case for `Sized` trait in `assembleCandidates` I can implement it in this PR - we already have `Ty.isSized` function and we use it in `assembleCandidates` directly for `Sized` trait case. What do you think about it?

vlad20012 replied:
Ok

Undin commented:
Oh, I've found out that these changes cause my favorite `IndexNotReadyException`

Undin commented:
* fixed `IndexNotReadyException`
* created special case in `assembleCandidates` for `Sized` trait
* changed run configuration to show `idea.log` in IDE

vlad20012 commented on src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt:
```diff
@@ -260,6 +272,21 @@ class ImplLookup(
+    private fun sizedTraitCandidates(ty: Ty, sizedTrait: RsTraitItem): Set<SelectionCandidate> {
+        var isSized = ty.isSized()
+        if (!isSized && ty is TyTypeParameter) {
+            isSized = ty.getTraitBoundsTransitively().any { it.element == sizedTrait }
```
Hmm, why we need this branch?

Undin replied:
In this case
```rust
trait Foo : Sized {
    fn foo(self: Self);
}
```
`Self` type should be `Sized` and we need to check super traits.
But constructing of `TyTypeParameter.Self` is better place to do it so I've moved corresponding code there.

vlad20012 replied:
Oh, Ok

vlad20012 commented:
bors r+

vlad20012 commented:
Thanks again :)
